feat(group/perm/sign): swap_adj_induction_on

zhangir-azerbayev opened the pull request:
1) Defines the multilinear map that sends a tuple of elements of an algebra to their product. 
2) ~~Defines the typeclass `alternating_map` that extends `multilinear_map`.~~ Superseded by #5102
3) Proves some lemmas about permutations and lists that are necessary for (1) and (2), and some lemmas about permutations of `fin n` that will be useful for formalizing exterior algebras. 
---
<!-- put comments you want to keep out of the PR commit here -->
This PR is essentially a number of lemmas and constructions I wrote for the `exterior_algebra` branch stated at the appropriate level of generality. 

A particular area where I could use some help is deciding where long proofs should be split into several shorter lemmas.

zhangir-azerbayev committed:
feat(data/list/basic): Added Mario's lemma pmap_map

zhangir-azerbayev committed:
added lemmas about permutations of fin n

zhangir-azerbayev committed:
feat(linear_algebra/multilinear): added the multinear algebra_prod

zhangir-azerbayev committed:
feat(linear_algebra/alternating): made linear_algebra/alternating

zhangir-azerbayev committed:
feat(group_theory/units_action): added units actions

zhangir-azerbayev committed:
feat(group_theory/perm/sign) added swap_induction_on'

zhangir-azerbayev committed:
feat(group_theory/group_action: removed units action

zhangir-azerbayev committed:
feat(linear_algebra/alternating): added map_perm

zhangir-azerbayev committed:
feat(linear_algebra/alternating: added map_perm

zhangir-azerbayev committed:
chore(linear_algebra/multilinear, linear_algebra/alternating): cleaned up proofs

zhangir-azerbayev committed:
doc(linear_algebra/alternating): added documentation

kim-em commented on src/data/list/basic.lean:
```diff
@@ -2327,6 +2327,10 @@ theorem map_pmap {p : α → Prop} (g : β → γ) (f : Π a, p a → β)
 by induction l; [refl, simp only [*, pmap, map]]; split; refl
 
+theorem pmap_map {α β γ} {p : β → Prop} (g : ∀ b, p b → γ) (f : α → β)
+  (l H) : pmap g (map f l) H = pmap (λ a h, g (f a) h) l (λ a h, H _ (mem_map_of_mem _ h)) :=
```
Could you add the types for `l` and `H`, for readability? (I know I'm being a pot-calling-a-kettle-black, but it is a good idea!)

kim-em replied:
Also, I think `α β γ` are already defined as variables, so you don't need them here.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -3,6 +3,7 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes
 -/
+import tactic
```
Please just import what you need.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -245,6 +246,26 @@ begin
+  suffices hQ : ∀ f, Q f,
+  { convert hQ f⁻¹,
+    simp,
+  },
```
Check the [style guide](https://leanprover-community.github.io/contribute/style.html): no braces-on-their-own-lines.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -245,6 +246,26 @@ begin
+lemma swap_induction_on' [fintype α] {P : perm α → Prop} (f : perm α) :
+  P 1 → (∀ f x y, x ≠ y → P f → P (f * swap x y)) → P f :=
+begin
+  intros h1 IH,
```
Just move arguments before the colon.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -245,6 +246,26 @@ begin
+  rw [mul_inv_rev, swap_inv],
+  apply IH,
+  exact h,
+  exact hQ,
```
Use braces whenever there are multiple goals open.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -693,4 +714,128 @@ using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ f, f.support
 
+lemma swap_eq_prod_swap_adj (n : ℕ) {i j : fin n} (hij : i ≠ j) : ∃ L : list (perm (fin n)),
+L.prod = equiv.swap i j ∧ ∀ l ∈ L, ∃ k : fin n, ∃ h : k.1 + 1 < n,
+ l = equiv.swap k ⟨k.1 + 1, h⟩ :=
```
This is quite a complicated statement. Would it be easier to say that there a list of nats, such that `equiv.swap i j` is the product of the corresponding adjacent transpositions?

kim-em replied:
Also, it might be easier if you split this into several lemmas: prove the first one with the hypothesis that `i < j`.

zhangir-azerbayev replied:
If I were to state the lemma using a list of nats, I can't think of a good way of keeping track of an `l.1 + 1 < n` for each `l ∈ L`.

kim-em replied:
You could, e.g. have `i j : fin (n+1)`, then produce in one declaration a `list (fin n)`, and then in a second declaration `equiv.swap i j = (L.map ...).prod`, where the ... converts a `fin n` into the adjacent transposition of `fin (n+1)`.

I'm not saying this is essential, just potentially results in cleaner statements, and proofs split into smaller chunks.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -693,4 +714,128 @@ using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ f, f.support
+L.prod = f ∧ ∀ l ∈ L, ∃ k : fin n, ∃ h : k.1 + 1 < n, l = equiv.swap k ⟨k.1 + 1, h⟩ :=
+begin
+  apply swap_induction_on f,
+  --Base case
```
Use braces.

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -693,4 +714,128 @@ using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ f, f.support
+  use add ++ base,
+  split,
+  {rw [←hbase.1, ←hadd.1],
+  rw list.prod_append},
```
Check style guide / mathlib on indenting

kim-em commented on src/group_theory/perm/sign.lean:
```diff
@@ -693,4 +714,128 @@ using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ f, f.support
+end
+
+lemma swap_adj_induction_on' {n : ℕ} {P : perm (fin n) → Prop} (f : equiv.perm $ fin n) :
+P 1 → (∀ f (x y : fin n), x.val + 1 = y.val → P f → P (f * swap x y)) → P f :=
```
Maybe don't reuse `f` here, for readability?

kim-em commented on src/linear_algebra/alternating.lean:
```diff
@@ -0,0 +1,115 @@
+alternating_map.map_alternating h hij
+
+lemma map_add_swap {i j : ι} (hij : i ≠ j):
+f ν + f (ν ∘ equiv.swap i j) = 0 :=
```
```suggestion
lemma map_add_swap {i j : ι} (hij : i ≠ j) :
  f ν + f (ν ∘ equiv.swap i j) = 0 :=
```

kim-em replied:
and similarly below.

kim-em commented on src/linear_algebra/alternating.lean:
```diff
@@ -0,0 +1,115 @@
+  convert key,
+  simp,
+  ext x,
+    cases classical.em (x = i) with hx hx,
```
Why the change in indenting here?

kim-em commented on src/linear_algebra/alternating.lean:
```diff
@@ -0,0 +1,115 @@
+class alternating_map extends multilinear_map R (λ i : ι , M) N :=
+(map_alternating {ν : ι → M} {i j : ι} (h : ν i = ν j) (hij : i ≠ j) : to_fun ν = 0)
+
+namespace alternating
```
Probably should be `alternating_map`.

kim-em commented on src/linear_algebra/alternating.lean:
```diff
@@ -0,0 +1,115 @@
+variables (N : Type*) [add_comm_group N] [module R N]
+variables (ι : Type*) [decidable_eq ι]
+
+class alternating_map extends multilinear_map R (λ i : ι , M) N :=
```
I don't think this should be a class? `multilinear_map` isn't.

zhangir-azerbayev committed:
style(linear_algebra/alternating): did reviewer suggestions for alternating.lean

zhangir-azerbayev committed:
style(data/list/basic) did reviewer suggestions

zhangir-azerbayev committed:
style(linear_algebra/multilinear): did reviewer suggestions

zhangir-azerbayev committed:
style(group_theory/perm/sign): reviewer suggestions

zhangir-azerbayev committed:
style(linear_algebra/multilinear): indenting

zhangir-azerbayev committed:
style(linear_algebra/multilinear): more indenting

zhangir-azerbayev committed:
style(linear_algebra/multilinear: even more indenting

zhangir-azerbayev committed:
style(group_theory/perm/sign): more reviewer suggestions

adamtopaz commented on src/linear_algebra/alternating.lean:
```diff
@@ -0,0 +1,112 @@
+permutation `sigma` of `ι`, we have that `f ν = (sign σ) f (ν ∘ σ)`.
+-/
+
+variables (R : Type*) [ring R]
```
The definition you have of `alternating_map` seems to be okay for semimodules over semirings.

zhangir-azerbayev replied:
Fixed

zhangir-azerbayev committed:
feat(linear_algebra/alternating): semimodules over semirings

zhangir-azerbayev committed:
fix(group_theory/perm/sign): removed unnecessary hypothesis from lemma

zhangir-azerbayev committed:
fix(linear_algebra/alternating, linear_algebra/multilinear): fixed linter complaints

Vierkantor commented on src/linear_algebra/multilinear.lean:
```diff
@@ -515,6 +516,135 @@ protected def pi_ring_equiv [fintype ι]  : M₂ ≃ₗ[R] (multilinear_map R (
+/--
+Auxiliary function used to define the multilinear map `algebra_prod`.
+-/
+def algebra_prod_aux : N := list.prod $ list.map ν (list.fin_range q)
```
It is often useful to use a general `fintype` rather than `fin` as an index for finite-dimensional vector spaces, so that would hypothetically look something like:
```suggestion
variable {q : Type*} [fintype q]
variable (ν : q → N)

@[priority 100] instance algebra_to_semimodule : semimodule S N := algebra.to_semimodule

/--
Auxiliary function used to define the multilinear map `algebra_prod`.
-/
def algebra_prod_aux : N := finset.univ.prod q
```
However, this would require that `N` is a `comm_semiring`, not just a `semiring`. Are there non-commutative examples of `N` that might be used? (As far as I know, exterior products are generally defined over a field, so my impression is that we can assume it is commutative.)

Vierkantor replied:
If the order of multiplication matters, maybe take the list of indices as a parameter:
```suggestion
variable {q : Type*} (qs : list q)
variable (ν : q → N)

@[priority 100] instance algebra_to_semimodule : semimodule S N := algebra.to_semimodule

/--
Auxiliary function used to define the multilinear map `algebra_prod`.
-/
def algebra_prod_aux : N := list.prod $ list.map ν qs
```
Apart from making the types more general, I expect that this can simplify the proofs below, because we can do the `fin_range` and the `prod` and `map` parts separately.

zhangir-azerbayev replied:
The non-commutative examples of N are the important ones, since the same file already deals with the commutative case.

zhangir-azerbayev replied:
If I do your second suggestion, can my proofs by induction on `q` be converted easily?

Vierkantor replied:
Then indeed I would use the second suggestion. You should be able to do the start off by induction on the list (instead of the size of the list that it was (arguably) before). For example, the start of `algebra_prod` might look like:

```lean
def algebra_prod : multilinear_map S (λ i : q, N) N :=
{ to_fun := algebra_prod_aux,
  map_add' :=
  begin
    intros ν i x y,
    induction qs with head tail hq,
```

zhangir-azerbayev replied:
I've been thinking about this some more and I have a couple of questions.
1. Instead of an index type and a list of indices, why not just use a `list N`? 
2. My understanding is that a `list` as implemented in Lean is a concept from functional programming, not from mathematics. If we define `algebra_prod` as taking a list, we are building a somewhat non-mathematical concept into our API, which makes me uneasy. From the perspective of a mathematician trying to understand mathlib, `fin q → N` makes a lot of sense since it's just a funny way of writing N^q, but the notion of a `list` as implemented in functional languages is not familiar. Is it not better to to hide everything with lists behind definitions?

zhangir-azerbayev replied:
@Vierkantor

Vierkantor replied:
Sorry for not responding, I was busy this weekend and didn't check my notifications enough.

1. If you think that works better, of course use it! I originally suggested a list of indices to make the analogy to the commutative `finset.sum` easier. I do not know if the rest might be easier for either way. (In my experience, I can only know that I made the right choices in a definition after I use it in proofs, before that I need to rely on intuition.)

2. I appreciate that the type `list N` is more "native" to functional programming than linear algebra, compared to `fin q -> N`. For the mathlib project, this is not an important consideration though. Contributions should use less beginner-friendly concepts if it makes the theory easier to work with once you learn them, and anyone is very much invited to ask questions to the community. (The (in)famous example is using filters to define limits of sequences and functions at once.) A different interface for the definitions can be useful, but I would say that would be part of future work, after this pull request.

I would like to make clear that I would still like to see the definition of `algebra_prod` and am already grateful for the original definition in terms of `fin q -> N`. Let's get it perfect before merging it, though.

eric-wieser replied:
With the push I just did, these `aux` definitions are completely gone, as they duplicate `multilinear_map.mk_pi_algebra_fin` from #4316

sgouezel commented on src/linear_algebra/multilinear.lean:
```diff
@@ -515,6 +516,135 @@ protected def pi_ring_equiv [fintype ι]  : M₂ ≃ₗ[R] (multilinear_map R (
+/--
+The multilinear map sending a vector of elements of a semiring to their product.
+-/
+def algebra_prod : multilinear_map S (λ i : fin q, N) N :=
```
This should make sense for any fintype, not only `fin q`.

eric-wieser replied:
This is precisely `multilinear_map.mk_pi_algebra_fin`, I think - so is no longer needed.

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into algebra_multilinear_maps

eric-wieser committed:
chore(*): Remove definitions which now exist elsewhere

For the sake of review, this leaves the definitions behind as aliases.
They will be removed in a later commit

eric-wieser committed:
chore(*): Fix proof broken by updates to fin in core lean

eric-wieser committed:
chore(*): Fix linter error

eric-wieser commented on src/group_theory/perm/sign.lean:
```diff
@@ -245,6 +246,24 @@ begin
+  apply IH,
+  { exact h,  },
+  { exact hQ, },
+end
```
Golfed to a single line in #5092, `λ h1 IH, inv_inv f ▸ swap_induction_on f⁻¹ h1 (λ f, IH f⁻¹)`

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into algebra_multilinear_maps

eric-wieser committed:
chore(group_theory/perm/sign): Golf a proof

eric-wieser committed:
chore(group_theory/perm/sign): Golf another proof

eric-wieser committed:
chore(group_theory/perm/sign): Golf another proof

eric-wieser committed:
Merge remote-tracking branch 'origin' into algebra_multilinear_maps

Note that this merge removes all the alternating map stuff from this branch and replaces it with the upstream version.
